src/main/java/com/google/cloud/spanner/pgadapter/parsers/ArrayParser.java

```diff
@@ -28,29 +32,66 @@
  * Translate wire protocol to array. Since arrays house any other specified types (including
  * potentially arrays), we use all parser types to parse each item within.
  */
-public class ArrayParser extends Parser<List> {
+public class ArrayParser extends Parser<List<?>> {
+  private static final String STRING_TOGGLE = "\"";
+  private static final String ARRAY_DELIMITER = ",";
+  private static final String PG_ARRAY_OPEN = "{", PG_ARRAY_CLOSE = "}";
+  private static final String SPANNER_ARRAY_OPEN = "[", SPANNER_ARRAY_CLOSE = "]";
 
-  private static String STRING_TOGGLE = "\"";
-  private static String ARRAY_DELIMITER = ",";
-  private static String PG_ARRAY_OPEN = "{", PG_ARRAY_CLOSE = "}";
-  private static String SPANNER_ARRAY_OPEN = "[", SPANNER_ARRAY_CLOSE = "]";
+  private final Type arrayElementType;
+  private final boolean isStringEquivalent;
 
-  int arrayType;
-  boolean isStringEquivalent;
-
-  public ArrayParser(ResultSet item, int position) throws SQLException {
+  public ArrayParser(ResultSet item, int position) {
     if (item != null) {
-      this.item = Arrays.asList((Object[]) item.getArray(position).getArray());
-      this.arrayType = item.getArray(position).getBaseType();
-      if (this.arrayType == Types.ARRAY) {
+      this.arrayElementType = item.getColumnType(position).getArrayElementType();
+      if (this.arrayElementType.getCode() == Code.ARRAY) {
         throw new IllegalArgumentException(
             "Spanner does not support embedded Arrays."
                 + " If you are seeing this, something went wrong!");
       }
-      this.isStringEquivalent = stringEquivalence(this.arrayType);
+      this.item = toList(item.getValue(position), this.arrayElementType.getCode());
+      this.isStringEquivalent = stringEquivalence(this.arrayElementType.getCode());
+    } else {
+      arrayElementType = null;
+      isStringEquivalent = false;
+    }
+  }
+
+  private List<?> toList(Value value, Code arrayElementType) {
+    switch (arrayElementType) {
+      case BOOL:
+        return value.getBoolArray();
+      case DATE:
+        return value.getDateArray();
+      case JSON:
+        return value.getJsonArray();
+      case BYTES:
+        return value.getBytesArray();
+      case INT64:
+        return value.getInt64Array();
+      case STRING:
+        return value.getStringArray();
+      case NUMERIC:
```

Comment:
should we error on non-PG types ?

Reply:
Yes, that seems better. Moved `NUMERIC` to the end to let it throw an error instead.